ipn/local.go

```diff
@@ -431,56 +445,105 @@ func (b *LocalBackend) updateFilter(netMap *controlclient.NetworkMap) {
 		b.e.SetFilter(filter.NewAllowNone(b.logf))
 		return
 	}
+	var data struct {
+		filter          filter.Matches
+		advertiseRoutes []wgcfg.CIDR
+		shieldsUp       bool
+	}
+	data.filter = netMap.PacketFilter
 
 	b.mu.Lock()
-	advRoutes := b.prefs.AdvertiseRoutes
+	data.advertiseRoutes = b.prefs.AdvertiseRoutes
+	data.shieldsUp = b.prefs.ShieldsUp
 	b.mu.Unlock()
-	localNets := wgCIDRsToFilter(netMap.Addresses, advRoutes)
 
-	if b.shieldsAreUp() {
+	changed := deepprint.UpdateHash(&b.filterHash, data)
+	if !changed {
+		b.logf("netmap packet filter: (not changed)")
+		return
+	}
+
+	localNets := wgCIDRsToFilter(netMap.Addresses, data.advertiseRoutes)
+
+	if data.shieldsUp {
 		// Shields up, block everything
 		b.logf("netmap packet filter: (shields up)")
 		var prevFilter *filter.Filter // don't reuse old filter state
 		b.e.SetFilter(filter.New(filter.Matches{}, localNets, prevFilter, b.logf))
 		return
 	}
 
-	// TODO(apenwarr): don't replace filter at all if unchanged.
-	// TODO(apenwarr): print a diff instead of full filter.
-	now := time.Now()
-	if now.Sub(b.lastFilterPrint) > 1*time.Minute {
-		b.logf("netmap packet filter: %v", netMap.PacketFilter)
-		b.lastFilterPrint = now
-	} else {
-		b.logf("netmap packet filter: (length %d)", len(netMap.PacketFilter))
-	}
+	b.logf("netmap packet filter: %v", netMap.PacketFilter)
 	b.e.SetFilter(filter.New(netMap.PacketFilter, localNets, b.e.GetFilter(), b.logf))
 }
 
+// dnsCIDRsEqual determines whether two CIDR lists are equal
+// for DNS map construction purposes (that is, only the first entry counts).
+func dnsCIDRsEqual(newAddr, oldAddr []wgcfg.CIDR) bool {
+	if len(newAddr) != len(oldAddr) {
+		return false
+	}
+	if len(newAddr) == 0 {
```

Comment:
could merge this into the final return expression:

    return len(newAddr) == 0  || newAddr[0] == oldAddr[0]